### DISCUSSION TITLE
Cannot run github action to publish python package

### DISCUSSION THREAD

**Discussion by ashwin-phadke** [community/community#25460] [Actions]
<p>I am trying to run <a href="https://github.com/ashwin-phadke/cvplayground/blob/54db6b75dbaa54eaf7aa4587424288429fa1a5f5/.github/workflows/python-publish.yml" rel="nofollow noopener">this</a> workflow to publish a python package from the directory <code>cvplay</code> repository inside the main repo <a href="https://github.com/ashwin-phadke/cvplayground" rel="nofollow noopener">cvplayground</a><br>
I have tried multiple answers here and on stackoverflow but keep getting</p>
<pre><code class="lang-auto">every step must define a `uses` or `run` key
</code></pre>
<p>or <code>setup.py</code> not found inside the directory and cannot figure out the main issue here.Need help</p>
<p>Here is the workflow file for quick view</p>
<pre><code class="lang-auto">
name: Upload Python Package

on: [push]

jobs:
  deploy:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: Change directory
        working-directory: ./cvplay
      - name: Install dependencies
        working-directory: ./cvplay
        run: |
          python -m pip install --upgrade pip
          pip install setuptools wheel twine
      - name: Build and publish
        working-directory: ./cvplay
        env:
          TWINE_USERNAME: ${{ secrets.TWINE_USERNAME }}
          TWINE_PASSWORD: ${{ secrets.TWINE_PASSWORD }}
        run: |
          python setup.py sdist bdist_wheel
          twine upload testpypi dist/* 
</code></pre>
<p>UPDATE :  Does it mean I add <code>uses</code> tag before every name?</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="ashwin-phadke" data-post="1" data-topic="130881">
<div class="title">
<div class="quote-controls"></div>
 ashwin-phadke:</div>
<blockquote>
<p>UPDATE : Does it mean I add <code>uses</code> tag before every name?</p>
</blockquote>
</aside>
<p>No, the error means each step must contain (order doesn’t matter) either a <code>uses</code> attribute (to use a predefined action) or a <code>run</code> attribute (to use a custom script).</p>
<p>There is one step in your workflow that doesn’t follow that rule:</p>
<aside class="quote no-group" data-username="ashwin-phadke" data-post="1" data-topic="130881">
<div class="title">
<div class="quote-controls"></div>
 ashwin-phadke:</div>
<blockquote>
<pre><code class="lang-yaml">      - name: Change directory
        working-directory: ./cvplay
</code></pre>
</blockquote>
</aside>
<p>You can just remove it.</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/ashwin-phadke">@ashwin-phadke</a>,</p>
<p>As the error mentioned, every step must define a <code>uses</code> or <code>run</code>  key.  However below step doens’t follow the rule:</p>
<pre><code class="lang-auto">      - name: Change directory
        working-directory: ./cvplay
</code></pre>
<p>You can remove step.  Instead of setting working directory for each step, you can set it in job level just once.</p>
<pre><code class="lang-auto">jobs:
  deploy:
    defaults:
      run:
        working-directory: ./cvplay   # job level working directory
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: Install dependencies
        run: |
          python -m pip install --upgrade pip
          pip install setuptools wheel twine
        ......
</code></pre>
<p>Doc here: <a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#defaultsrun" rel="nofollow noopener">https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#defaultsrun</a></p>
<p>Thanks</p>

**Comment by ashwin-phadke** [comment]
<p>Thank you <a class="mention" href="/weide-zhou">@weide-zhou</a> , that helped, only one thing gives error now after I specified python versions using <code>setup-python</code> that twine command not found. If all other packages are being installed by default in <code>home</code> directory how can one package install in some other place, I guess command not found was more pertaining to this error message that I get while installing dependencies</p>
<p><code>WARNING: The script wheel is installed in '/home/runner/.local/bin' which is not on PATH.   Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.</code> multiple times.</p>
<p>Here is the error I get in actual build and publish part :<br>
<code>/home/runner/work/_temp/ace6d6e16507.sh: line 2: twine: command not found ##[error]Process completed with exit code 127. </code></p>
<p>and this is how I have specified python version in the workflow :</p>
<pre><code class="lang-auto">      uses: actions/setup-python@v2
        with:
          python-version: '3.x'
      - name: Install dependencies
</code></pre>
<p>and I do</p>
<pre><code class="lang-auto">        run: |
          python -m pip install --upgrade pip
          pip3 install --user setuptools wheel twine
</code></pre>
<p>to install packages, I added --user option to overcome this error but it is still there.</p>
<p>Here is the full updated workflow</p>
<pre><code class="lang-auto"># This workflows will upload a Python Package using Twine when a release is created
# For more information see: https://help.github.com/en/actions/language-and-framework-guides/using-python-with-github-actions#publishing-to-package-registries

name: Upload Python Package

on: [push]

jobs:
  deploy:
    defaults:
      run:
        working-directory: ./cvplay   # job level working directory
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: Set up Python
        uses: actions/setup-python@v2
        with:
          python-version: '3.x'
      - name: Install dependencies
        run: |
          python -m pip install --upgrade pip
          pip3 install --user setuptools wheel twine
      - name: Build and publish
        env:
          TWINE_USERNAME: ${{ secrets.TWINE_USERNAME }}
          TWINE_PASSWORD: ${{ secrets.TWINE_PASSWORD }}
        run: |
          python setup.py sdist bdist_wheel
          twine upload testpypi dist/* 

</code></pre>
<p>UPDATE :  adding <code>~/.local/bin/twine</code> worked but there should be a workaround</p>

**Comment by ashwin-phadke** [comment]
<p>Thank you that helped.</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/ashwin-phadke">@ashwin-phadke</a>,</p>
<p>By default, the twine is installed in below location and runner can find it without issue:</p>
<pre><code class="lang-auto">pip3 install setuptools wheel twine
</code></pre>
<p><img src="https://user-images.githubusercontent.com/52772761/181091215-cf153a40-765b-442a-8377-21c4ab67ef57.png" alt="image" data-base62-sha1="j9s16F7wioagjr0qxAytunKz2dJ" width="405" height="109"></p>
<p>If you install in user directory, system cannot find <code>twine</code> command automatically, and you need add system path to fix it.</p>
<pre><code class="lang-auto">echo "::add-path::/home/runner/.local/bin"
</code></pre>
<p><img src="https://user-images.githubusercontent.com/52772761/181091216-0d35d70e-6a02-495e-bad5-64f579c9cd9b.png" alt="image" data-base62-sha1="roPNWr1dcwGuOLH12AicD5JuWIX" width="389" height="116"></p>
<p>Doc here: <a href="https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#adding-a-system-path" rel="nofollow noopener">https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#adding-a-system-path</a></p>
<p>Thanks</p>

**Comment by ashwin-phadke** [comment]
<p>For me it is by default getting installed at /local/bin.<br>
Thanks for the help.</p>

**Comment by jadhavmanoj** [comment]
<p>I used docker image. I installed the python packages successfully. But when I ran <code>python manage.py runserver</code> command it says package django not found.</p>
<p>ModuleNotFoundError: No module named ‘django’ or rather not finding installed packages.</p>
